### PR TITLE
Binary domain: boundary conditions and cleanup

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

- Quite a bit of cleanup and simplifications in the tests. This was largely motivated by me wanting to make adding the boundary conditions easier for myself
- Remove the equiangular option since the resulting domain is incorrect
- Add boundary conditions to the binary domain
- Some clang-tidy stuff, yay

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### CODE REVIEW

File: tests/Unit/Domain/Creators/Test_BinaryCompactObject.cpp
Comment (nilsvu) [inline tests/Unit/Domain/Creators/Test_BinaryCompactObject.cpp:314]: This commit is just formatting, did you intend to make more changes in this commit? Else please reword the commit message.

File: src/Domain/Creators/BinaryCompactObject.hpp
Comment (nilsvu) [inline src/Domain/Creators/BinaryCompactObject.hpp]: I think you meant to name this "InnerBoundary", and the one below "OuterBoundary", and put them in a `BoundaryConditionsGroup`, like you did for `Interval` (I like the grouping you did there).

**Review by nilsvu — CHANGES_REQUESTED** [review]

**Comment by nilsdeppe** [comment]
I made the requested changes and pushed a new commit that replaces the error tests with `CHECK_THROWS` tests. Reducing the number of test cases will make running the unit tests faster, which, while not the bottleneck in CI, means people are more likely to run the tests locally.

File: src/Domain/Creators/BinaryCompactObject.cpp
Comment (fmahebert) [inline src/Domain/Creators/BinaryCompactObject.cpp:170]: If I don't excise A or B (say for a BNS run) then what inner boundary condition do I want to pass? None? Or should we not require the inner BC in this scenario?

File: src/Domain/Creators/BinaryCompactObject.hpp
Comment (fmahebert) [inline src/Domain/Creators/BinaryCompactObject.hpp]: For consistency with Interval et al, return "InnerBoundary" here ... or actually maybe "ExcisionBoundaries" is even better, to make clear it'll apply to any/all excision regions if present?

File: src/Domain/Creators/BinaryCompactObject.hpp
Comment (fmahebert) [inline src/Domain/Creators/BinaryCompactObject.hpp:276]: To make clearer this applied to all excision regions, I suggest one of,
- append a sentence like "This boundary condition will be applied to all inner excision boundaries", OR
- "inner" -> "excision region"

File: src/Domain/Creators/BinaryCompactObject.hpp
Comment (fmahebert) [inline src/Domain/Creators/BinaryCompactObject.hpp]: For consistency with Interval et al, return "OuterBoundary" here

File: src/Domain/Creators/BinaryCompactObject.cpp
Comment (nilsdeppe) [reply src/Domain/Creators/BinaryCompactObject.cpp:170]: I don't know how to not require a BC in this case because we would effectively need a "default" value. The only ways to really handle this that I can think of right now are:
1. have an `Alternatives<tmpl::list<ExciseA, ExciseB, InnerBoundaryCondition>, tmpl::list<>>`, but I don't think this will work because this is effectively setting a "default" value
2. leave as-is with a comment
3. decide to add a None boundary condition, either system-specific or general. Probably a general one like the periodic would work

Comment (kidder) [reply src/Domain/Creators/BinaryCompactObject.cpp:170]: My suggestion is to:  
- Add a `None` BC that errors if it is ever used
- Either require that `None` is used if ExciseInteriorA and ExciseInteriorB are both false; or add a comment to InnerBoundaryCondition's help string saying that this option is ignored in that case

Comment (fmahebert) [reply src/Domain/Creators/BinaryCompactObject.cpp:170]: I would be satisfied with a comment (as per @kidder's suggestion) for now, and perhaps adding a `None` BC in a separate PR (or making an issue for it).

**Comment by nilsdeppe** [comment]
Dependent on #2884

**Comment by nilsdeppe** [comment]
I have squashed the previous fixup and pushed a new one that use `None` to do error checking when nothing is excised.

I did give `Options::Auto<...>` a try for not having this semi-redundant options, but this is non-trivial because `std::optional<std::unique_ptr>` behaves differently than `std::unique_ptr` in terms of implicit conversions to base class pointers. If someone wants to give that a try, I have no objections to using that design, but I can't really spend time figuring that out right now.

File: src/Domain/Creators/BinaryCompactObject.cpp
Comment (kidder) [inline src/Domain/Creators/BinaryCompactObject.cpp]: I would say "Inner boundary condition must be None if ExciseInteriorA and ExciseInteriorB are both false" to directly cite the input file options.  If you do so, don't forget to update the corresponding text in the `Catch::Matchers::Contains`

**Review by kidder — COMMENTED** [review]
just one small thing that can be squashed when @fmahebert and @nilsleiffischer also give the go ahead to squash

File: src/Domain/Creators/BinaryCompactObject.cpp
Comment (fmahebert) [reply src/Domain/Creators/BinaryCompactObject.cpp]: I like this suggestion 👍 

I would also suggest moving this `if` block down below the `if` blocks on lines 166–180. This has the aesthetic appeal of checking for simple mistakes before checking for complicated ones... and the code appeal that below line 166–173 you no longer need the innermost "inner == nullptr" check because you know "(outer == nullptr) == (inner == nullptr)" and you're in an `if` block where "outer != nullptr", which implies "inner != nullptr".

File: src/Domain/Creators/BinaryCompactObject.cpp
Comment (fmahebert) [inline src/Domain/Creators/BinaryCompactObject.cpp]: Option to write this as "(outer == nullptr) != (inner == nullptr)" instead... totally up to you I'm happy either way

**Review by fmahebert — COMMENTED** [review]
You have my permission to squash, including these changes if any.

File: src/Domain/Creators/BinaryCompactObject.cpp
Comment (nilsdeppe) [reply src/Domain/Creators/BinaryCompactObject.cpp]: I find that much harder to understand, honestly :)

File: src/Domain/Creators/BinaryCompactObject.cpp
Comment (nilsdeppe) [reply src/Domain/Creators/BinaryCompactObject.cpp]: While I mostly agree, then I need to rework a lot of logic in the test cases for what different iteration of the for loop triggers what error, and that turns out to be quite tedious and time consuming

File: src/Domain/Creators/BinaryCompactObject.cpp
Comment (nilsdeppe) [reply src/Domain/Creators/BinaryCompactObject.cpp]: Turns out there's a thing for this: `xor`. I've pushed a separate fixup so it's easily dropped if people don't like it

**Review by kidder — DISMISSED** [review]

File: src/Domain/Creators/BinaryCompactObject.cpp
Comment (fmahebert) [reply src/Domain/Creators/BinaryCompactObject.cpp]: Oh yeah 😅

**Comment by fmahebert** [comment]
LGTM

**Comment by kidder** [comment]
@nilsleiffischer please look at the fixups

**Review by nilsvu — COMMENTED** [review]
LGTM 👍 Please don't wait for my re-approval after this gets squashed.

**Comment by nilsdeppe** [comment]
Rebased on develop and squashed in the fixups. I also squashed in `inner_bc_ == nullptr xor outer_bc_ == nullptr` -> `(inner_bc_ == nullptr) xor (outer_bc_ == nullptr)` because of a GCC warning

**Review by fmahebert — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Comment by kidder** [comment]
@nilsleiffischer this has been squashed

**Review by nilsvu — APPROVED** [review]

**Pull request merged**
